src/controllers/userController.ts

```diff
@@ -0,0 +1,78 @@
+import { Request, Response, NextFunction } from 'express';
+import { UserService } from '../services/userService';
+import jwt from 'jsonwebtoken';
+import { randomUUID } from 'crypto';
+import { BadRequest, Unauthorized } from '../error/errors';
+import Joi from 'joi';
+import { users, User } from '../storage/storage';
+import { hashPassword, comparePassword } from '../utils/bcrypt';
+
+
+const userService = new UserService();
+
+const userSchema = Joi.object({
+  email: Joi.string().email().required().messages({
+    'string.email': 'Invalid email format',
+    'any.required': 'Email is required',
+  }),
+  name: Joi.string().required().messages({
+    'any.required': 'Name is required',
+  }),
+  password: Joi.string()
+    .min(8)
+    .pattern(new RegExp('^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)(?=.*[?!@#$%^&*])[A-Za-z\\d?!@#$%^&*]{8,}$'))
+    .required()
+    .messages({
+      'string.min': 'Password must be at least 8 characters long',
+      'string.pattern.base': 'Password must include upper and lower case letters, numbers, and special characters',
+      'any.required': 'Password is required',
+    }),
+});
+
+export const registerUser = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { email, name, password } = req.body;
+    const { error } = userSchema.validate({ email, name, password });
+
+    if (error) {
+      throw new BadRequest(error.details[0].message);
+    }
+
+    const passwordHash = await hashPassword(password);  // Хешування пароля
+    const newUser: User = { id: randomUUID(), email, name, passwordHash, role: 'CUSTOMER' };
```

Comment:
Create `enum` to describe the user roles
[https://www.typescriptlang.org/docs/handbook/enums.html](https://www.typescriptlang.org/docs/handbook/enums.html)